### PR TITLE
Increase the default message size limit of the grpc protocol for device manager

### PR DESCRIPTION
**Pull request by DrAuYueng**
#### What type of PR is this?

/kind cleanup
/kind regression

#### What this PR does / why we need it:
In our environment graphics card resource specifications on different nodes in the same cluster vary greatly. Some nodes have smaller graphics card memory such as 2G which are used for program logic verification testing, and some are larger such as 80G which are used for model training. In order to make full use of these resources, we divide the graphics card into (MiB/GiB) units based on the device plugin mechanism and schedule them uniformly. When the unit is GiB, it works well. When the unit is MiB and the node resource exceeds a certain value, such as 240G (80G*3), the grpc communication between kubelet and device plugin is abnormal. The abnormal log is as follows: 
```
"listAndWatch ended unexpectedly for device plugin" err="rpc error: code = ResourceExhausted desc = grpc: received message larger than max (14958030 vs. 4194304)" resource="foo.com/gpu-mem"
```
Current `grpc.MaxCallRecvMsgSize` default is 4MB(`4*1024*1024`=4194304), we think it is necessary to adjust the value of grpc.MaxCallRecvMsgSize to `16*1024*1024`, reference: https://github.com/kubernetes/kubernetes/blob/df577d7fbc4c76b4b660007565fae86b48a1086f/staging/src/k8s.io/cri-client/pkg/remote_runtime.go#L97.

#### Which issue(s) this PR fixes:

Fixes #

#### Does this PR introduce a user-facing change?

```release-note
Increase the default message size limit of the grpc protocol for device manager
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs

```

### CODE REVIEW

**Comment by kannon92** [comment]
Please add a release note.

/remove-kind regression cleanup
/kind feature

**Comment by kannon92** [comment]
/triage accepted
/priority backlog

The change seems reasonable to me. WDYT @ffromani  @bart0sh?

**Comment by DrAuYueng** [comment]
It looks like the failed unit tests are not relevant to this PR.

**Comment by ffromani** [comment]
/lgtm

I can't see any issue either; we use (IIRC) the same value for podresources API so makes sense

/assign @klueska 
to doublecheck there is no ill side effect (I can't think of any either but Kevin has surely more experience about devices)

**Comment by DrAuYueng** [comment]
@klueska Could you take a look at this PR？Thanks.

**Comment by SergeyKanzhelev** [comment]
> we divide the graphics card into (MiB/GiB) units based on the device plugin mechanism and schedule them uniformly.

How many devices are you getting this way? I am surprised 4MB is not enough and wonder if anything else may break with so many devices? Or the IDs of devices are so large?

If the number of devices is so big, let's add a test into https://github.com/kubernetes/kubernetes/blob/master/test/e2e_node/device_plugin_failures_test.go to make sure that things are generally working with this number of devices, there may be other sideeffects of this. 

One example I am worried right away is in https://github.com/kubernetes/enhancements/issues/4680 we can reach some limits on the pod status size when adding health of all devices to the pod status.

**Comment by DrAuYueng** [comment]
> Or the IDs of devices are so large?

Yes, there are so many device IDs, these are virtual IDs, which actually map to a real physical device.

**Comment by SergeyKanzhelev** [comment]
> > Or the IDs of devices are so large?
> 
> Yes, there are so many device IDs, these are virtual IDs, which actually map to a real physical device.

So can you please add a e2e test for that many devices? I expect many things may break, not just message size

**Comment by ffromani** [comment]
> > > Or the IDs of devices are so large?
> > 
> > 
> > Yes, there are so many device IDs, these are virtual IDs, which actually map to a real physical device.
> 
> So can you please add a e2e test for that many devices? I expect many things may break, not just message size

good point. The ideal setup would be a reproduction of your setup @DrAuYueng , with of course device details anonymized if needed. It would be super nice to have comments/rationale about from where the devices come from, to make it clear this is a real-world reproducer and not a made-up scenario to break the limits.

**Comment by DrAuYueng** [comment]
> > > Or the IDs of devices are so large?
> > 
> > 
> > Yes, there are so many device IDs, these are virtual IDs, which actually map to a real physical device.
> 
> So can you please add a e2e test for that many devices? I expect many things may break, not just message size

Sure.

File: test/e2e_node/device_plugin_failures_test.go [hunk 1]
```diff
@@ -352,4 +352,61 @@ var _ = SIGDescribe("Device Plugin Failures:", framework.WithNodeConformance(),
 		// after the graceful period devices capacity will reset to zero
 		gomega.Eventually(getNodeResourceValues, devicePluginGracefulTimeout+1*time.Minute, f.Timeouts.Poll).WithContext(ctx).WithArguments(resourceName).Should(gomega.Equal(ResourceValue{Allocatable: 0, Capacity: 0}))
 	})
+
+	ginkgo.It("when ListAndWatch will fail immediately in shared GPU scenario, node allocatable will be set to zero and kubelet will not retry to list resources", func(ctx context.Context) {
```
Comment (SergeyKanzhelev) [inline test/e2e_node/device_plugin_failures_test.go:356]: is this ready for review? Looks like WIP

Comment (DrAuYueng) [reply test/e2e_node/device_plugin_failures_test.go]: Ready now.

File: pkg/kubelet/cm/devicemanager/plugin/v1beta1/api.go [hunk 2]
```diff
@@ -46,4 +46,7 @@ const (
 	errInvalidResourceName = "the ResourceName %q is invalid"
 	// errBadSocket is the error raised when the registry socket path is not absolute
 	errBadSocket = "bad socketPath, must be an absolute path:"
+	// maxMsgSize use 16MB as the default message size limit.
+	// grpc library default is 4MB
+	maxMsgSize = 1024 * 1024 * 16
```
Comment (SergeyKanzhelev) [inline pkg/kubelet/cm/devicemanager/plugin/v1beta1/api.go:51]: I am not sure how it works on gRPC. Will Device Plugin need to change this size too? Or only one side can configure it to a higher value?

File: pkg/kubelet/cm/devicemanager/plugin/v1beta1/api.go [hunk 3]
```diff
@@ -46,4 +46,7 @@ const (
 	errInvalidResourceName = "the ResourceName %q is invalid"
 	// errBadSocket is the error raised when the registry socket path is not absolute
 	errBadSocket = "bad socketPath, must be an absolute path:"
+	// maxMsgSize use 16MB as the default message size limit.
```
Comment (SergeyKanzhelev) [inline pkg/kubelet/cm/devicemanager/plugin/v1beta1/api.go:49]: the comment is kind of just explaining code. A better comment will be: 

```suggestion
	// maxMsgSize use 16MB as the message size limit for Device Plugin communication to support cases of XXX devices configured by the Device Plugin.
```

File: test/e2e_node/device_plugin_failures_test.go [hunk 4]
```diff
@@ -352,4 +352,61 @@ var _ = SIGDescribe("Device Plugin Failures:", framework.WithNodeConformance(),
 		// after the graceful period devices capacity will reset to zero
 		gomega.Eventually(getNodeResourceValues, devicePluginGracefulTimeout+1*time.Minute, f.Timeouts.Poll).WithContext(ctx).WithArguments(resourceName).Should(gomega.Equal(ResourceValue{Allocatable: 0, Capacity: 0}))
 	})
+
+	ginkgo.It("when ListAndWatch will fail immediately in shared GPU scenario, with a large number of devices node allocatable will be set to zero and kubelet will not retry to list resources", func(ctx context.Context) {
+		// randomizing so tests can run in parallel
+		resourceName := fmt.Sprintf("test.device/%s", f.UniqueName)
+		devicePrefix := "GPU-e9ebf26b-6a2c-634b-f6e6-3552ffafa7d8"
+		// 10 GPUs, 80G each, converted to MiB, used to simulate a large number of devices
+		total := 10 * 80 * 1024
```
Comment (SergeyKanzhelev) [inline test/e2e_node/device_plugin_failures_test.go:361]: Add a comment saying that this number of devices exceeds the max message size limit

File: pkg/kubelet/cm/devicemanager/plugin/v1beta1/api.go (hunk 3 above)
Comment (SergeyKanzhelev) [reply pkg/kubelet/cm/devicemanager/plugin/v1beta1/api.go:49]: Also add a small note on why this is a reasonable number of devices (how one can achieve it in real world scenario)

File: test/e2e_node/device_plugin_failures_test.go [hunk 5]
```diff
@@ -352,4 +352,61 @@ var _ = SIGDescribe("Device Plugin Failures:", framework.WithNodeConformance(),
...
+		gomega.Eventually(getNodeResourceValues, nodeStatusUpdateTimeout, f.Timeouts.Poll).WithContext(ctx).WithArguments(resourceName).Should(gomega.Equal(ResourceValue{Allocatable: 0, Capacity: 0}))
+
+		// kubelet will never retry ListAndWatch (this will sleep for a long time)
+		gomega.Consistently(plugin.Calls, devicePluginUpdateTimeout, f.Timeouts.Poll).Should(gomega.HaveLen(2))
+
+		// however kubelet will not delete the resource
+		gomega.Eventually(getNodeResourceValues, nodeStatusUpdateTimeout, f.Timeouts.Poll).WithContext(ctx).WithArguments(resourceName).Should(gomega.Equal(ResourceValue{Allocatable: 0, Capacity: 0}))
+	})
+
+	ginkgo.It("will allocatable to a number of healthy devices in shared GPU scenario", func(ctx context.Context) {
```
Comment (SergeyKanzhelev) [inline test/e2e_node/device_plugin_failures_test.go:387]:
```suggestion
	ginkgo.It("when number of devices is large, but within supported limit (in shared GPU scenario), will allocatable to a number of healthy devices", func(ctx context.Context) {
```

File: test/e2e_node/device_plugin_failures_test.go [hunk 6]
```diff
@@ -352,4 +352,61 @@ var _ = SIGDescribe("Device Plugin Failures:", framework.WithNodeConformance(),
...
+		// randomizing so tests can run in parallel
+		resourceName := fmt.Sprintf("test.device/%s", f.UniqueName)
+		devicePrefix := "GPU-33054273-1377-182c-ab17-a9c1b2834745"
+		// a GPU with 80G memory, converted to MiB
+		total := 1 * 80 * 1024
```
Comment (SergeyKanzhelev) [inline test/e2e_node/device_plugin_failures_test.go:392]: so how one will consume this? That many Containers will be scheduled? The limit for # of pods is typically 128, maybe 256. How many of those shared devices will be consumed by each container and pod?

Asking because I am trying to think what else will break

**Comment by SergeyKanzhelev** [comment]
@johnbelamaric is this a supported case for DRA? I wonder what will break on such partitioned node in DRA case

**Comment by SergeyKanzhelev** [comment]
A few more things to test:

1. Do PodResources API work with that many devices?
2. Can status of those devices successfully be stored in https://github.com/kubernetes/enhancements/issues/4680?
3. Will the channel in pkg/kubelet/cm/devicemanager/manager.go (with the buffer of 100) be sufficient here?

I am still unsure about the motivation to split GPUs in such a small pieces. Can you tell more about your scenario?

**Comment by johnbelamaric** [comment]
How many devices are we talking about?

Whether DRA would support this depends on the driver author - they driver author certainly could create an oversubscription model like this.

I think we probably need some overall limit on # of devices per driver per node. I don't know what a reasonable number is, but we will want to test the scalability of the scheduler as well as plugin mechanisms, and object sizes as we approach those limits. For DRA, I am 99.9% sure that 10x80x1024 devices on a node will absolutely crush the scheduler...especially if you have 1000 of those nodes!

We have limits in the ResourceSlice API on a per-slice basis, but it is possible to use multiple slices to represent devices on a single node, so those aren't necessarily truly limiting. There are limits on all the ResourceClaim values, but those apply only to a single claim, not to all claims on a node. Similarly, for resource health in pod status, that only reports items associated with the specific *pod*.

**Comment by klueska** [comment]
The k8s-device-plugin for GPUs allows one to specify a set of "replicas" for a GPU, which allows it to be oversubscribed for use by up to `numReplicas` users. Beecause of the way the device-plugin API is designed, we must advertise `numReplicas` distinct devices to the kubelet (as a list), which it in in turn uses to populate the number of `nvidia.com/gpu` resources available on the node.

It sounds like @DrAuYueng is either leveraging this support directly, or some custom variant of it that allows the GPU memory to be partitioned in addition to oversubscribing the GPUs. In any case, the underlying requirement to advertise each "replica" as a distinct device remains unchanged.

Just as with the device plugin, DRA requires that driver authors list each allocatable device separately.  So if you wanted to implement such a "system-mediated" sharing scheme that mimics what the device-plugin is doing, you would run into the problems @johnbelamaric describes above.

However, the typical use-case for sharing GPUs with DRA is *not* via "system-mediated" sharing, but rather "user-mediated" sharing. Meaning that you use DRA to bind a full device to a ResourceClaim, and then you share / oversubscribe it among different containers / pods by referencing that shared device at the application layer. Meaning you don't rely on the system to mediate the sharing of the device for you -- instead the user controls precisely how that device get's shared among its clients *after* it has been bound to a claim.

For example, to achieve the equivalent of what @DrAuYueng describes in the description of this PR would be achieved by the following with DRA:
```
---
apiVersion: resource.k8s.io/v1alpha3
kind: ResourceClaim
metadata:
  name: shared-gpu-1Mi
spec:
  devices:
    requests:
    - name: gpu
      deviceClassName: gpu.nvidia.com
    config:
      opaque:
        driver: gpu.nvidia.com
        parameters:
          apiVersion: gpu.nvidia.com/v1alpha1
          kind: GpuConfig
          sharing:
            strategy: MPS
            mpsConfig:
              defaultPinnedDeviceMemoryLimit: 1Mi
```

With this in place, every container (or more accurately every CUDA client) that references the `gpu` device in the `shared-gpu-1Mi` will get access to exactly `1Mi` of GPU memory.

So in summary, while you *could* run into similar issues with DRA, you shouldn't actually in practice, because the sharing model is different.

**Comment by SergeyKanzhelev** [comment]
@DrAuYueng can you please see if DRA will help your scenario better?

I am very hesitant to increase the number of devices we support to such a big number. It will make things too fragile even if we will be able to resolve all the things we can guess would potentially break.

If the scenario you have in mind is not addressable by DRA, let's talk about this scenario a little bit more

**Comment by SergeyKanzhelev** [comment]
/close

as we need way more investigation before changing this limit
